src/collector.rs

```diff
@@ -0,0 +1,388 @@
+                scope.defer(move || a.into_owned());
+
+                assert!(!(*scope.bag).is_empty());
+
+                while !(*scope.bag).is_empty() {
+                    scope.flush();
+                }
+            });
+        }
+    }
+
+    #[test]
+    fn garbage_buffering() {
+        let collector = Collector::new();
+        let handle = collector.handle();
+        drop(collector);
+
+        handle.pin(|scope| unsafe {
+            for _ in 0..10 {
+                let a = Owned::new(7).into_ptr(scope);
+                scope.defer(move || a.into_owned());
+            }
+            assert!(!(*scope.bag).is_empty());
+        });
+    }
+
+    #[test]
+    fn pin_holds_advance() {
+        let collector = Collector::new();
+
+        let threads = (0..NUM_THREADS)
+            .map(|_| {
+                scoped::scope(|scope| {
+                    scope.spawn(|| {
+                        let handle = collector.handle();
+                        for _ in 0..500_000 {
+                            handle.pin(|scope| {
+                                let before = collector.0.get_epoch();
+                                collector.0.collect(scope);
+                                let after = collector.0.get_epoch();
+
+                                assert!(after.wrapping_sub(before) <= 2);
+                            });
+                        }
+                    })
+                })
+            })
+            .collect::<Vec<_>>();
+        drop(collector);
+
+        for t in threads {
+            t.join();
+        }
+    }
+
+    #[test]
+    fn incremental() {
+        const COUNT: usize = 100_000;
+        static DESTROYS: AtomicUsize = ATOMIC_USIZE_INIT;
+
+        let collector = Collector::new();
+        let handle = collector.handle();
+
+        handle.pin(|scope| unsafe {
+            for _ in 0..COUNT {
+                let a = Owned::new(7i32).into_ptr(scope);
+                scope.defer(move || {
+                    drop(a.into_owned());
+                    DESTROYS.fetch_add(1, Ordering::Relaxed);
+                });
+            }
+            scope.flush();
+        });
+
+        let mut last = 0;
+
+        while last < COUNT {
+            let curr = DESTROYS.load(Ordering::Relaxed);
+            assert!(curr - last <= 1024);
+            last = curr;
+
+            handle.pin(|scope| collector.0.collect(scope));
+        }
+        assert!(DESTROYS.load(Ordering::Relaxed) == 100_000);
+    }
+
+    #[test]
+    fn buffering() {
+        const COUNT: usize = 10;
+        static DESTROYS: AtomicUsize = ATOMIC_USIZE_INIT;
+
+        let collector = Collector::new();
+        let handle = collector.handle();
+
+        handle.pin(|scope| unsafe {
+            for _ in 0..COUNT {
+                let a = Owned::new(7i32).into_ptr(scope).as_raw() as usize;
+                scope.defer(move || {
+                    drop(Owned::from_raw(a as *mut i32));
```

Comment:
I think we can use `a.into_owned()` here, and remove the `.as_raw() as usize` part above.